### ISSUE TITLE
Heuristics for clustered units

### ISSUE THREAD

**Issue by jvoigts** [SpikeInterface/spiketoolkit#283]
I'd like to check what you all are thinking about adding a mechanism for, or repository of, heuristics to decide if a cluster returned by a spike sorter is of neural origin. This could be a set of measures and associated thresholds that are applied in postprocessing.

It seems that quite a few users of automated sorters use ad-hoc scripts, often in addition to manual curation, in order to mark and remove clusters that are clearly artifacts. For instance, by checking if the ISI distribution is sensible, and if the spike waveform looks neural etc.

As a 'off-label' use of these heuristics, one could also add classifications, such as regular vs. fast spiking cortical neurons etc.

These heuristics are of course somewhat dependent on electrode type, brain region, and experimental setup, etc. but there should be enough commonalities to make sharing these heuristics useful.

If you think this should go somewhere else, I'd also be interested to hear where, and what form such a repository could take.

**Comment by mhhennig** [comment]
This is a very good question, and yes, we're looking a this issue.

As you say, looking at spike train features can be a little dangerous as one cannot be sure what to expect from neurons (except perhaps ISI violations). A possible starting point are the quality metrics. We have not done this systematically, but a ground truth data analysis could perhaps reveal how to use these.

If you are interested, you can find and download a large ground truth data set from the spikeforest project (https://spikeforest.flatironinstitute.org/) - they can be obtained as SI extractors directly, very handy.

Let us know how you get on - we'll keep this issue open to collect info on the matter.
Matthias

**Comment by alejoe91** [comment]
Hi @jvoigts!

> I'd like to check what you all are thinking about adding a mechanism for, or repository of, heuristics to decide if a cluster returned by a spike sorter is of neural origin. This could be a set of measures and associated thresholds that are applied in postprocessing.

I believe @jsiegle is working on something like this. It could be unsupervised based on quality metrics (see validation and curation modules), or even supevised at one point.

> It seems that quite a few users of automated sorters use ad-hoc scripts, often in addition to manual curation, in order to mark and remove clusters that are clearly artifacts. For instance, by checking if the ISI distribution is sensible, and if the spike waveform looks neural etc.
> As a 'off-label' use of these heuristics, one could also add classifications, such as regular vs. fast spiking cortical neurons etc.

Following this https://github.com/SpikeInterface/spiketoolkit/issues/255 I wanted to add a `FeatureCallculator` in the postprocessing module. The postprocessing module could also be used for RS-FS separation based on waveforms, ISI, etc. I think it would fit perfectly.

Another thing that I wanted to add (and have some code for it), is automatic tracking of units over sessions. I think that would be a great postprocessing feature to have.

> These heuristics are of course somewhat dependent on electrode type, brain region, and experimental setup, etc. but there should be enough commonalities to make sharing these heuristics useful.
> 
> If you think this should go somewhere else, I'd also be interested to hear where, and what form such a repository could take.

I think all of this fits perfectly in spiketoolkit, mostly in the `postprocessing` and `curation` modules. 

I'll make an initial PR with a skeleton of the `FeatureCalculator`. 
Regarding the automatic noise cluster recognition we can probably hear what @jsiegle thinks :)

Cheers
Alessio

**Comment by samuelgarcia** [comment]
+1

**Comment by jsiegle** [comment]
I agree this is super important. You can't objectively compare sorter performance without first filtering out artifactual units...but of course the characteristics of these units will vary from sorter to sorter, so there will not be a one-size-fits-all approach to identifying them.

We have developed a [Python module](https://github.com/AllenInstitute/ecephys_spike_sorting/tree/master/ecephys_spike_sorting/modules/noise_templates) that does this reasonably well for the combination of Neuropixels + Kilosort2. But to make it more general will require some careful thought. I think the best approach would be to design a class that accepts mean waveforms and spike times as input, then outputs a label for each unit. Then there can be sub-classes that implement the classification, whether it's via a set of heuristics, a neural network, or something else. SpikeToolkit can provide a set of standard classifiers, while also making it easy to insert custom code. Hopefully over time we can build up a library of classifiers that are optimized for different situations!

We also have a module for [calculating waveform-based metrics](https://github.com/AllenInstitute/ecephys_spike_sorting/tree/master/ecephys_spike_sorting/modules/mean_waveforms) that can be used for FS vs. RS distinctions (or [RS1 vs. RS2](https://www.physiology.org/doi/abs/10.1152/jn.00680.2018)). This could be combined with the waveform features that help with ID'ing noise units, but it may be better to keep these metrics separate, since the things you want to know about real neurons are often different from what's helpful for making the neuron vs. non-neuron distinction.

**Comment by alejoe91** [comment]
@jsiegle very cool! I guess we can start integrating those if that's allright for you :)

@jvoigts I haven't had time to make the PR. I'll try over this week to get us started.

**Comment by samuelgarcia** [comment]
@jsiegle : I think we can indeed compare sorters performence **wihout** cleaning. The result give you an important idea of how much we need to clean or not the result of a sorter.

If SI provide some good tools for cleaning and level weakness of sorters then it is a good thing.

But comparing the raw output of sorters is still an important results for the community.

**Comment by jsiegle** [comment]
@alejoe91, sure, go ahead! But both modules assume high-density, equally spaced recording sites, so they will have to be substantially revised to work with more types of electrodes.

@samuelgarcia I agree there is value to comparing the raw outputs of each sorter, but you still need some way to classify noise vs. neural units in order to judge how much post-hoc cleaning is required.

**Comment by colehurwitz** [comment]
@samuelgarcia i agree with both you and Josh. The best way to do this analysis is before and after cleaning. Let’s try to do this for the paper if feasible.
